packages/vite/src/node/plugins/resolve.ts

```diff
@@ -651,95 +702,202 @@ export function tryNodeResolve(
+      }
+
+      try {
+        resolvedId = resolver(usedId, resolvedPkg, targetWeb, options)
+        if (resolvedId) {
+          break
+        }
+      } catch (err) {
+        if (!options.tryEsmOnly) {
+          throw err
+        }
+      }
+      if (options.tryEsmOnly) {
+        resolvedId = resolver(usedId, resolvedPkg, targetWeb, {
+          ...options,
+          isRequire: false,
+          mainFields: DEFAULT_MAIN_FIELDS,
+          extensions: DEFAULT_EXTENSIONS
+        })
+        if (resolvedId) {
+          break
         }
       }
+
+      // Reset the resolvedPkg variables to avoid false positives as we
+      // continue our search.
+      resolvedPkg = undefined
+      resolvedPkgId = undefined
+      resolvedPkgType = undefined
+      continue
+    }
+
+    // No package.json was found, but there could still be a module
+    // here. To match Node's behavior, we must be able to resolve a
+    // module without a package.json file helping us out.
+    try {
+      const stat = fs.statSync(entryPath)
+      if (stat.isFile()) {
+        resolvedId = entryPath
+        break
+      }
+      resolvedId = tryIndexFile(entryPath, targetWeb, options)
+      if (resolvedId) {
+        break
+      }
+    } catch {}
+
+    // In case a file extension is missing, we need to try calling the
+    // `tryFsResolve` function.
+    let entryDir = path.dirname(entryPath)
+    let entryDirExists = false
+    if (entryDir === nodeModulesDir) {
+      entryDirExists = true
+    } else {
+      try {
+        const stat = fs.statSync(entryDir)
+        entryDirExists = stat.isDirectory()
+      } catch {}
+    }
+
+    if (entryDirExists) {
+      resolvedId = tryFsResolve(
+        entryPath,
+        { ...options, skipPackageJson: true },
+        false,
+        targetWeb
+      )
+      if (resolvedId) {
+        break
+      }
     }
-    return
-  }
 
-  let resolveId = resolvePackageEntry
-  let unresolvedId = pkgId
-  const isDeepImport = unresolvedId !== nestedPath
-  if (isDeepImport) {
-    resolveId = resolveDeepImport
-    unresolvedId = '.' + nestedPath.slice(pkgId.length)
+    // Stop looking if we're at the workspace root directory.
+    if (
+      isWorkspaceRoot(
+        path.dirname(nodeModulesDir),
+        preserveSymlinks,
+        packageCache
+      )
+    )
+      break
   }
 
-  let resolved: string | undefined
-  try {
-    resolved = resolveId(unresolvedId, pkg, targetWeb, options)
-  } catch (err) {
-    if (!options.tryEsmOnly) {
-      throw err
+  if (!resolvedId) {
+    const mayBeOptionalPeerDep =
+      !options.isHookNodeResolve &&
+      basedir !== root &&
+      !isBuiltin(basePkgId) &&
+      !basePkgId.includes('\0') &&
+      bareImportRE.test(basePkgId)
+
+    if (!mayBeOptionalPeerDep) {
+      return // Module not found.
+    }
+
+    // Find the importer's nearest package.json with a "name" field.
+    // Some projects (like Svelte) have nameless package.json files to
```

Comment:
this will probably change in the future with Svelte 4 when we drop support for old versions of Node, so worth including the version here
```suggestion
    // Some projects (like Svelte 3) have nameless package.json files to
```